### PR TITLE
Add exception handling

### LINKED ISSUE #6650: Loading invalid save crashes the game/falsely say it's loaded
**OS:** Windows 10
**Version:** 0.1.2
**Commit/Build:** 891de0a

When loading a save that is invalid, no error message pops up and you stay in the current state you were at, with a marker on the loaded save which indicates that it's loaded.
In some cases depending on what the original extension were - crashes the game
**Not reproducible in vanilla** (gives out "contains invalid data" error tooltip

- [ ] Reproducible in RCT2 (vanilla)?
- [ ] Multiplayer?

**Steps to reproduce:**
1. Create a new text-document in the save folder
2. Change extension to .sv6
3. Load the file in-game

--------------------------

Dumpfile from loading a .txt converted to .sv6
**Dumpfile from commit 891de0a created by task-manager, examined**

Error message/missing file to create full callstack: 
Could not locate: d:\rs1\minkernel\crts\ucrt\src\appcrt\startup\abort.cpp

Callstack;
```
>	openrct2.exe!abort() Line 77	C++	Symbols loaded.
 	openrct2.exe!terminate() Line 58	C++	Symbols loaded.
 	[External Code]		Annotated Frame
 	openrct2.dll!FileStream::Read(void * buffer, unsigned __int64 length) Line 179	C++	Symbols loaded.
 	openrct2.dll!IStream::ReadArray<unsigned char>(unsigned __int64 count) Line 101	C++	Symbols loaded.
 	[Inline Frame] openrct2.dll!TryClassifyAsS4(IStream *) Line 100	C++	Symbols loaded.
 	openrct2.dll!TryClassifyFile(IStream * stream, ClassifiedFileInfo * result) Line 56	C++	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::LoadParkFromStream(IStream * stream, const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & path, bool loadTitleScreenFirstOnFail) Line 389	C++	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::LoadParkFromFile(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & path, bool loadTitleScreenOnFail) Line 383	C++	Symbols loaded.
 	openrct2.dll!context_load_park_from_file(const char * path) Line 880	C++	Symbols loaded.
 	[Inline Frame] openrct2.dll!window_loadsave_invoke_callback(int) Line 737	C++	Symbols loaded.
 	openrct2.dll!window_loadsave_select(rct_window * path, const char *) Line 778	C++	Symbols loaded.
 	openrct2.dll!window_loadsave_scrollmousedown(rct_window * w, int scrollIndex, int x, int y) Line 412	C++	Symbols loaded.
 	openrct2.dll!input_widget_left(int x, int y, rct_window * w, short widgetIndex) Line 1010	C	Symbols loaded.
 	openrct2.dll!game_handle_input_mouse(int x, int y, int state) Line 297	C	Symbols loaded.
 	openrct2.dll!game_handle_input(...) Line 129	C	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::Update() Line 769	C++	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::RunVariableFrame() Line 730	C++	Symbols loaded.
 	[Inline Frame] openrct2.dll!OpenRCT2::Context::RunFrame() Line 665	C++	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::RunGameLoop() Line 640	C++	Symbols loaded.
 	openrct2.dll!OpenRCT2::Context::RunOpenRCT2(int argc, const char * * argv) Line 160	C++	Symbols loaded.
 	openrct2.dll!NormalisedMain(int argc, const char * * argv) Line 62	C++	Symbols loaded.
 	openrct2.dll!LaunchOpenRCT2(int argc, wchar_t * * argvW) Line 53	C++	Symbols loaded.
 	[External Code]		Annotated Frame
```

Local on pos1
```
		sigabrt_action	Variable is optimized away and not available.	
```
Local on pos4
```
+		this	0x0000006ffd0ff088 {_file=0x000001f371df35e0 {_Placeholder=0x0000000000000000 } _ownsFilePtr=true _canRead=...}	FileStream *
		buffer	0x000001f37d3f9680	void *
		length	0	unsigned __int64
		remainingBytes	0	unsigned __int64
```

### PR DESCRIPTION
**Pull request by Chaosmeister**
Add missing exception handling in
* TryClassifyAsS4
* TryClassifyAsTD4_TD6
* LoadParkFromStream

Fix #6650 - Loading invalid save crashes the game/falsely say it's loaded

### CODE REVIEW

**Comment by IntelOrca** [comment]
If an exception occurs you may get a memory leak. The pointers need to be changed to `unique_ptr` or freed in the catch.

File: src/openrct2/FileClassifier.cpp [hunk 1]
```diff
@@ -95,57 +95,68 @@ static bool TryClassifyAsS6(IStream * stream, ClassifiedFileInfo * result)
...
+        }
+        else if (type == FILE_TYPE_SC4)
+        {
+            result->Type = FILE_TYPE::SCENARIO;
+            result->Version = version;
+            success = true;
+        }
     }
-    else if (type == FILE_TYPE_SC4)
+    catch (Exception)
```
Comment (rwjuk) [inline src/openrct2/FileClassifier.cpp]: We should at very least log the error message (as in the other handler you added).

File: src/openrct2/FileClassifier.cpp [hunk 2]
```diff
@@ -95,57 +95,68 @@ static bool TryClassifyAsS6(IStream * stream, ClassifiedFileInfo * result)
...
+                    result->Type = FILE_TYPE::TRACK_DESIGN;
+                    result->Version = version;
+                    success = true;
+                }
             }
         }
-        Memory::Free(td6data);
     }
-    Memory::Free(data);
+    catch (Exception)
```
Comment (rwjuk) [inline src/openrct2/FileClassifier.cpp]: Same here - log the error to the console.

**Comment by IntelOrca** [comment]
@rwjuk Are you happy with this now?

**Comment by Gymnasiast** [comment]
Bump @rwjuk

**Comment by rwjuk** [comment]
Whoops, sorry, missed this one. Yes.

**Pull request merged**